### PR TITLE
IGNITE-14916 Calcite. RuntimeHashIndex scan must filter result by predicate

### PR DESCRIPTION
**Pull request by tledkov-gridgain**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RuntimeHashIndex.java [hunk 1]
```diff
@@ -108,11 +109,15 @@ private GroupKey key(Row r) {
...
         private final Supplier<Row> searchRow;
 
+        /** Row filter. */
+        private final Predicate<Row> filter;
+
         /**
          * @param searchRow Search row.
          */
-        IndexScan(Supplier<Row> searchRow) {
+        IndexScan(Supplier<Row> searchRow, Predicate<Row> filter) {
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RuntimeHashIndex.java:118]: Let's mark filter as `@Nullable` (here and up in the stack)

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/HashIndexSpoolExecutionTest.java [hunk 2]
```diff
@@ -115,6 +115,7 @@ public void testIndexSpool() throws Exception {
                     ctx,
                     rowType,
                     ImmutableBitSet.of(0),
+                    null,
```
Comment (alex-plekhanov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/HashIndexSpoolExecutionTest.java:118]: Let's use `testFilter` and add some test case with filtering, for example, something like:
```
                        new GridTuple3(
                            (Predicate<Object[]>)(r -> ((int)r[2]) == 0),
                            new Object[] {size / 2 + 1, null, null},
                            1
                        )
```

**Review by alex-plekhanov — COMMENTED** [review]
Additionally:
- Condition modification and TODO with ticket number should be cleaned up from `FilterSpoolMergeToHashIndexSpoolRule`
- JavaDoc for `IgniteHashIndexSpool.cond` field about selectivity should be changed
- Serialization/deserialization for condition should be added to `IgniteHashIndexSpool.explainTerms`/`IgniteHashIndexSpool(RelInput input)`

**Comment by tledkov-gridgain** [comment]
@alex-plekhanov , thanks for review.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RuntimeHashIndex.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RuntimeHashIndex.java:119]: Still not nullable

**Review by alex-plekhanov — APPROVED** [review]

**Pull request merged**